### PR TITLE
net/gcoap: some more minor improvements

### PR DESCRIPTION
**Pull request by haukepetersen**
- added defines for request codes -> as I run into trouble when using nanocoap's codes (which are bitfields and not enum values)
- improved the doxygen for some `gcoap_resp..` functions
- added/fixed some debug output

### CODE REVIEW

File: sys/include/net/gcoap.h [hunk 1]
```diff
@@ -241,6 +241,16 @@ extern "C" {
 #define GCOAP_PAYLOAD_MARKER    (0xFF)
 
 /**
+ * @name    Possible request codes
```
Comment (smlng) [inline sys/include/net/gcoap.h]: aren't they called _methods_?

File: sys/include/net/gcoap.h [hunk 2]
```diff
@@ -241,6 +241,16 @@ extern "C" {
...
+ * @{
+ */
+#define GCOAP_GET               (1) /**< GET request */
```
Comment (smlng) [inline sys/include/net/gcoap.h]: maybe use anonymous enum, instead?

File: sys/include/net/gcoap.h [hunk 3]
```diff
@@ -412,11 +422,11 @@ void gcoap_register_listener(gcoap_listener_t *listener);
 /**
  * @brief  Initializes a CoAP request PDU on a buffer.
  *
- * @param[in] pdu Request metadata
- * @param[in] buf Buffer containing the PDU
- * @param[in] len Length of the buffer
- * @param[in] code Request code
- * @param[in] path Resource path
+ * @param[in] pdu   Request metadata
```
Comment (smlng) [inline sys/include/net/gcoap.h]: should be `[in,out]`?

File: sys/include/net/gcoap.h [hunk 4]
```diff
@@ -430,10 +440,10 @@ int gcoap_req_init(coap_pkt_t *pdu, uint8_t *buf, size_t len, unsigned code,
  * Assumes the PDU has been initialized with gcoap_req_init() or
  * gcoap_resp_init().
  *
- * @param[in] pdu Request metadata
- * @param[in] payload_len Length of the payload, or 0 if none
- * @param[in] format Format code for the payload; use COAP_FORMAT_NONE if not
- *                   specified
+ * @param[in] pdu           Request metadata
```
Comment (smlng) [inline sys/include/net/gcoap.h]: dito, `[in,out]`

File: sys/include/net/gcoap.h [hunk 5]
```diff
@@ -443,11 +453,11 @@ ssize_t gcoap_finish(coap_pkt_t *pdu, size_t payload_len, unsigned format);
 /**
  *  @brief Writes a complete CoAP request PDU when there is not a payload.
  *
- * @param[in] pdu Request metadata
- * @param[in] buf Buffer containing the PDU
- * @param[in] len Length of the buffer
- * @param[in] code Request code
- * @param[in] path Resource path
+ * @param[in] pdu   Request metadata
```
Comment (smlng) [inline sys/include/net/gcoap.h]: dito

File: sys/net/application_layer/coap/gcoap.c [hunk 6]
```diff
@@ -302,7 +302,7 @@ static void _find_resource(coap_pkt_t *pdu, coap_resource_t **resource_ptr,
 static ssize_t _finish_pdu(coap_pkt_t *pdu, uint8_t *buf, size_t len)
 {
     ssize_t hdr_len = _write_options(pdu, buf, len);
-    DEBUG("gcoap: header length: %u\n", hdr_len);
+    DEBUG("gcoap: header length: %i\n", (int)hdr_len);
```
Comment (smlng) [inline sys/net/application_layer/coap/gcoap.c]: actually I think the error handling in `_write_options` is _suboptimal_. I mean the function actually returns `size_t` but also some negative error code, plus this cast its _ugly_. IMHO verify that the url starts with `/` should be done somewhere else.

File: sys/net/application_layer/coap/gcoap.c [hunk 7]
```diff
@@ -452,6 +452,7 @@ static ssize_t _write_options(coap_pkt_t *pdu, uint8_t *buf, size_t len)
         size_t url_len = strlen((char *)pdu->url);
         if (url_len) {
             if (pdu->url[0] != '/') {
```
Comment (smlng) [inline sys/net/application_layer/coap/gcoap.c]: hä, this is redundant, because already done in `_write_options`, though I oppose that, too (see above).

mhm, maybe leave as is and fix only the methods code and docu stuff. And fix this url checking in a separate PR?

File: sys/include/net/gcoap.h (hunk 2 above)
Comment (kaspar030) [inline sys/include/net/gcoap.h]: And what exactly ist the problem with the corresponding nanocoap defines?

File: sys/net/application_layer/coap/gcoap.c (hunk 6 above)
Comment (haukepetersen) [inline sys/net/application_layer/coap/gcoap.c]: I agree, but this is a separate issue, this PR simply fixes the output of the signed variable...

File: sys/net/application_layer/coap/gcoap.c (hunk 7 above)
Comment (haukepetersen) [inline sys/net/application_layer/coap/gcoap.c]: same here, I agree with the issue, though this is not related to this PR...

File: sys/include/net/gcoap.h (hunk 1 above)
Comment (haukepetersen) [inline sys/include/net/gcoap.h]: true, the RFC calls them `CoAP Method Codes`, will fix

File: sys/include/net/gcoap.h (hunk 2 above)
Comment (haukepetersen) [inline sys/include/net/gcoap.h]:
```c
#define GCOAP_GET               (1) /**< GET request */
#define GCOAP_POST              (2) /**< POST request */
#define GCOAP_PUT               (3) /**< PUT request */
#define GCOAP_DELETE            (4) /**< DELETE request */
```
vs.
```c
#define COAP_GET                (0x1)
#define COAP_POST               (0x2)
#define COAP_PUT                (0x4)
#define COAP_DELETE             (0x8)
```

Comment (haukepetersen) [inline sys/include/net/gcoap.h]: oh, but I just see, that nanocoap also defines:
```
#define COAP_CLASS_REQ          (0)
#define COAP_METHOD_GET         (1)
#define COAP_METHOD_POST        (2)
#define COAP_METHOD_PUT         (3)
#define COAP_METHOD_DELETE      (4)
```
So these should do the trick, though very confusing from a user point of view...

Comment (kaspar030) [inline sys/include/net/gcoap.h]: (there' "COAP_METHOD_GET/PUT/POST' in addition to the bitfield values.

**Comment by kaspar030** [comment]
> though very confusing from a user point of view...

True that. One set is used for the headers (defined by protocol), the other set is used for bitfields (in order for the handlers to be able to do ```COAP_GET | COAP_POST```. Happy to change define names, but both is needed.

**Comment by haukepetersen** [comment]
I agree (and understand) that both are needed, we just need to improve the documentation (and possibly the naming) to make it more intuitive to use for noobs like me :-)

**Comment by kb2ma** [comment]
Or maybe use NANOCOAP_HANDLER_GET OR NANOCOAP_RESOURCE_GET for the bitfield names to indicate that they are specific to nanocoap rather than CoAP itself.

**Comment by haukepetersen** [comment]
- removed the `GCOAP_xxx` definitions again
- fixed many style/consistency/indention/[in,out] issues in the `gcoap.h` header

File: sys/net/application_layer/coap/gcoap.c [hunk 8]
```diff
@@ -452,6 +452,7 @@ static ssize_t _write_options(coap_pkt_t *pdu, uint8_t *buf, size_t len)
...
+                DEBUG("gcoap: _write_ptions: path does not start with '/'\n");
```
Comment (kb2ma) [inline sys/net/application_layer/coap/gcoap.c]: Spelling of "_write_options".

**Comment by haukepetersen** [comment]
fixed typo

**Comment by haukepetersen** [comment]
@kaspar030 are you happy with this PR?

**Comment by smlng** [comment]
need to resolve conflicts

**Comment by haukepetersen** [comment]
done

**Comment by haukepetersen** [comment]
and go.
